viz: support `stat` on new policy resources

Pothulapati opened the pull request:
Fixes #6733

As policy resources provide a grouping, statistics summaries should
also be allowed on these groupings which are useful to the user. Them
being port specific provide a great way to break down these metrics
further.

This PR adds support for policy resources i.e `server` and `serverauthorization`
 on the `stat` command.

## Changes

This adds a new path in the `stat_summary.go` file to handle policy
objects. I tried to see if we could re-use some of the other paths
but some of the labels seems to differ and hence a different path
had to be created. We can try to refactor and merge them though.

We support both request and TCP metrics for the `server` resource
while only the former with `serverauthorization` resources
as metrics are generated in this manner.

This also adds these policy objects into the `k8s` package to
make them as known resources.

For both the policy resources, `--from` doesn't work as these
metrics are not exposed from outbound, and there is no way to
query about the client workload from the inbound metrics. `--to`
is supported to get metrics specifically for a destination workload.
(just like on a service)

## Testing

```bash
> curl -sL https://run.linkerd.io/emojivoto.yml | linkerd inject --proxy-log-level debug - | kubectl apply -f -

> kubectl apply -f https://gist.githubusercontent.com/Pothulapati/69edb59fb1c54c3d993e1ad5145571d4/raw/897de1a8d57e3606420ca845c6abfd69bf241fb2/emojivoto-policy.yml

> linkerd  viz stat srv/emoji-grpc -n emojivoto -owide
NAME         MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN   READ_BYTES/SEC   WRITE_BYTES/SEC
emoji-grpc        -   100.00%   2.0rps           1ms           1ms           1ms          1         199.9B/s         2208.0B/s

> linkerd  viz stat srv/web-http -n emojivoto -owide
NAME       MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN   READ_BYTES/SEC   WRITE_BYTES/SEC
web-http        -    94.02%   1.9rps           4ms           9ms          10ms          1         152.7B/s         4505.9B/s

> linkerd  viz stat srv -n emojivoto -o wide                                                      
NAME          MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN   READ_BYTES/SEC   WRITE_BYTES/SEC
emoji-grpc         -   100.00%   2.0rps           1ms           1ms           1ms          1         201.6B/s         2209.8B/s
prom               -         -        -             -             -             -          -                -                 -
voting-grpc        -    86.21%   1.0rps           1ms           1ms           1ms          1          98.3B/s           55.9B/s
web-http           -    91.67%   2.0rps           3ms           8ms          10ms          1         157.7B/s         4600.3B/s


> linkerd  viz stat serverauthorization/web-public -n emojivoto
NAME       MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99  
web-http        -    89.83%   2.0rps           3ms           9ms          10ms

> linkerd viz stat saz -n emojivoto
NAME          AUTHORIZATION     MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
emoji-grpc    emoji-grpc             -   100.00%   2.0rps           1ms           1ms           1ms
prom          prom-prometheus        -         -        -             -             -             -
voting-grpc   voting-grpc            -    89.83%   1.0rps           1ms           1ms           1ms
web-http      web-public             -    94.96%   2.0rps           1ms           5ms           9ms

> linkerd viz stat saz/web-public -n emojivoto                                                 
NAME       AUTHORIZATION   MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
web-http   web-public           -    90.00%   2.0rps           1ms           5ms           9ms
```

Signed-off-by: Tarun Pothulapati <tarunpothulapati@outlook.com>

<!--  Thanks for sending a pull request!

If you already have a well-structured git commit message, chances are GitHub
set the title and description of this PR to the git commit message subject and
body, respectively. If so, you may delete these instructions and submit your PR.

If this is your first time, please read our contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md

The title and description of your Pull Request should match the git commit
subject and body, respectively. Git commit messages are structured as follows:

```
Subject

Problem

Solution

Validation

Fixes #[GitHub issue ID]

DCO Sign off
```

Example git commit message:

```
Introduce Pull Request Template

GitHub's community guidelines recommend a pull request template, the repo was
lacking one.

Introduce a `PULL_REQUEST_TEMPLATE.md` file.

Once merged, the
[Community profile checklist](https://github.com/linkerd/linkerd2/community)
should indicate the repo now provides a pull request template.

Fixes #3321

Signed-off-by: Jane Smith <jane.smith@example.com>
```

Note the git commit message subject becomes the pull request title.

For more details around git commits, see the section on Committing in our
contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md#committing
-->

alpeb commented on viz/metrics-api/stat_summary_test.go:
```diff
@@ -1400,6 +1400,50 @@ status:
+											},
+										},
+									},
+								},
```
So these unit tests will get completed with srv and saz objects in the `k8sConfigs` fixtures?

alpeb commented on viz/metrics-api/stat_summary.go:
```diff
@@ -366,6 +368,52 @@ func (s *grpcServer) trafficSplitResourceQuery(ctx context.Context, req *pb.Stat
+		}
+	}
+
+	keys := make([]rKey, 0)
+	for k := range requestMetrics {
+		keys = append(keys, k)
+	}
+	for k := range tcpMetrics {
+		keys = append(keys, k)
+	}
```
This is the part still pending that you mentioned, where k8s objects need to be filtered?

alpeb replied:
Also, the lack of info about the k8s objects results in tables with no names, like:
```console
$ bin/go-run cli viz stat server -n emojivoto
NAME   MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN
            -   100.00%   1.2rps           1ms           1ms           1ms          4
```

Pothulapati replied:
Yep, That's the reason. As we are directly querying the prometheus metrics.

alpeb commented:
I think you forgot to add RBAC to `metrics-api` to query servers?
```
$ bin/linkerd viz stat -n emojivoto -owide server/emoji-grpc 
StatSummary API error: servers.policy.linkerd.io "emoji-grpc" is forbidden: User "system:serviceaccount:linkerd-viz:metrics-api" cannot get resource "servers" in API group "policy.linkerd.io" in the namespace "emojivoto"
```

alpeb commented:
Also would be awesome if you issued a PR against the emojivoto repo with the policies you're using for testing :wink:

Pothulapati commented:
> I think you forgot to add RBAC to metrics-api to query servers?

Yep, Missed doing that. Was running the `metrics-api` server locally (which uses the k8s client svc account) when testing

> Also would be awesome if you issued a PR against the emojivoto repo with the policies you're using for testing wink

Sure, Will do that.

Pothulapati commented:
Added support to breakdown `saz` queries  into servers. Also, added support for `[UNAUTHORIZED]` when you query for all `saz` resources.

```bash
> linkerd viz stat saz -n emojivoto
NAME          AUTHORIZATION     MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
emoji-grpc    [UNAUTHORIZED]         -     0.00%   2.0rps           0ms           0ms           0ms
emoji-grpc    emoji-grpc             -   100.00%   2.0rps           1ms           1ms           1ms
prom          [UNAUTHORIZED]         -     0.00%   0.0rps           0ms           0ms           0ms
prom          prom-prometheus        -         -        -             -             -             -
voting-grpc   [UNAUTHORIZED]         -     0.00%   1.0rps           0ms           0ms           0ms
voting-grpc   voting-grpc            -    89.83%   1.0rps           1ms           1ms           1ms
web-http      [UNAUTHORIZED]         -     0.00%   1.9rps           0ms           0ms           0ms
web-http      web-public             -    94.96%   2.0rps           1ms           5ms           9ms


```

Pothulapati commented on viz/metrics-api/stat_summary.go:
```diff
@@ -897,10 +915,80 @@ func (s *grpcServer) getPolicyMetrics(ctx context.Context, req *pb.StatSummaryRe
 }
 
+func processPolicyPrometheusMetrics(results []promResult, groupBy model.LabelNames) (map[sazKey]*pb.BasicStats, map[sazKey]*pb.TcpStats) {
+	basicStats := make(map[sazKey]*pb.BasicStats)
```
Tried not to repeat this but using a different key meant that we had to do this and different key had to be used not only to differentiate `Server` and `ServerAuthorization` resources but also to have a different `metricToKey` func which was important to get the queries and results right.

I'm sure this can be simplified but it could be a significant change (which I think we should definitely take up after this)

olix0r commented:
I'm seeing some unexpected output:

```
:; linkerd viz stat -n emojivoto saz
NAME          AUTHORIZATION    MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
emoji-grpc    [UNAUTHORIZED]        -     0.00%   1.2rps           0ms           0ms           0ms
emoji-grpc    emoji-grpc            -   100.00%   1.2rps           1ms           1ms           1ms
voting-grpc   [UNAUTHORIZED]        -     0.00%   0.6rps           0ms           0ms           0ms
voting-grpc   voting-grpc           -    88.57%   0.6rps           1ms           1ms           1ms
web-http      [UNAUTHORIZED]        -     0.00%   1.2rps           0ms           0ms           0ms
web-http      web-public            -    97.18%   1.2rps           3ms           5ms           5ms
```

The RPS of the `[unauthorized]` row is always the same as the authorized row...

olix0r commented:
Also, it's not really helpful/relevant to view latency or success rate for unauthorized requests. we only ever care about RPS.

I also don't think it's relevant to display a MESHED column in srv or saz views...

adleong commented:
It's a bit confusing that when you do a query for serverauthorizations, the `NAME` column is the name of the server, not the name of the saz:

```console
> bin/linkerd viz stat saz -n emojivoto
NAME   AUTHORIZATION   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
grpc   mesh-mtls        93.30%   3.0rps           1ms           1ms           1ms
http   mesh-mtls        91.67%   2.0rps           2ms           3ms           3ms
```

It would be more clear if the name column referred to the requested resource (i.e. the saz) and instead there was a column for the server:

```console
> bin/linkerd viz stat saz -n emojivoto
NAME        SERVER   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
mesh-mtls   grpc        93.30%   3.0rps           1ms           1ms           1ms
mesh-mtls   http        91.67%   2.0rps           2ms           3ms           3ms
```

But perhaps even better would be to remove this extra column entirely and JUST give the user the stats for the resource they asked for.  i.e.

```console
> bin/linkerd viz stat saz -n emojivoto
NAME        SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
mesh-mtls   93.30%   3.0rps           1ms           1ms           1ms
```

(with potentially the addition of a "Unauthorized RPS" column for servers.  question: should the Unauthorized column only be included in wide or json outputs?)

```console
> bin/linkerd viz stat server -n emojivoto
NAME   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   UNAUTHORIZED
grpc   93.30%   3.0rps           1ms           1ms           1ms    1.7rps
http   91.67%   2.0rps           2ms           3ms           3ms    0.5rps
```

I think that this make the behavior more consistent with how `stat` works for other resource types and should simplify the code since we would no longer needed to handle server authorization keys differently in the metrics processing code.

adleong commented:
If we follow my suggestion above, it does leave us without the ability to break down by server and saz at the same time.  However, I think that's the perfect situation for a command which works similarly to `linkerd authz`.  I'd recommend (in a follow-up PR) creating a new command: `linkerd viz auth <resource>`

```console
> bin/linkerd viz authz -n emojivoto deploy/web
SERVER   AUTHORIZATION   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
grpc     mesh-mtls        93.30%   3.0rps           1ms           1ms           1ms
grpc     [UNAUTHORIZED]       -    0.5rps             -             -             -
http     mesh-mtls        91.67%   2.0rps           2ms           3ms           3ms
http     [UNAUTHORIZED]       -    1.1rps             -             -             -
```

This can be thought of like the `linkerd auth` command with stats attached to each row.

Pothulapati commented:
Okay, So Updated this PR to work similarly to other resources, and not do anything extra for `stat saz`.

Will be working on the following things next in separate PRs:

- Update Server view to include `unauthorized rps` column
- Add new `viz authz` that shows  the `saz X srv` view in a single table.

adleong reviewed (commented):
This looks great!  Strangely, metrics seem to be missing when querying for "all namespaces":

```
alex@dev:~/workspace/linkerd2$ bin/linkerd viz stat saz -A
NAMESPACE   NAME        MESHED   SUCCESS   RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
emojivoto   mesh-mtls        -         -     -             -             -
alex@dev:~/workspace/linkerd2$ bin/linkerd viz stat saz -n emojivoto
NAME        MESHED   SUCCESS   RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
mesh-mtls   93.96%    5.0rps   1ms           2ms           5ms
```

adleong commented:
Also, in my example above the columns appear to be misaligned with the success rate in the MESHED column and so on.

Pothulapati commented on viz/metrics-api/stat_summary.go:
```diff
@@ -366,6 +398,89 @@ func (s *grpcServer) trafficSplitResourceQuery(ctx context.Context, req *pb.Stat
 	return resourceResult{res: &rsp, err: nil}
 }
 
+func (s *grpcServer) getUnstructuredResources(req *pb.StatSummaryRequest) ([]rKey, error) {
```
Makes sense

Pothulapati commented:
Looks like just reverting the changes wasn't enough, and caused some weird printing things. So, Updated this PR to display results as expected.

Pothulapati commented:
Added a new `unauthorized` field that has the RPS of the denied requests.

```bash
# Initial values
on ⛵ kind-kind  linkerd2 on 🌱 taru [📦📝🤷‍] via 🐼 v1.16.7 via  via ❄️  impure (shell)
➜ ./bin/go-run cli viz stat srv -A -owide                                                                                                         ~/work/linkerd2
NAMESPACE   NAME          UNAUTHORIZED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN   READ_BYTES/SEC   WRITE_BYTES/SEC
emojivoto   emoji-grpc          0.0rps   100.00%   1.8rps           1ms           1ms           3ms          1         188.6B/s         2072.9B/s
emojivoto   prom                0.0rps         -        -             -             -             -          -                -                 -
emojivoto   voting-grpc         0.0rps    80.70%   0.9rps           1ms           2ms           3ms          1          91.4B/s           52.7B/s
emojivoto   web-http            0.0rps    90.68%   2.0rps           2ms          10ms          28ms          1         153.7B/s         4509.4B/s

# After changing the `emoji-grpc` authz
on ⛵ kind-kind  linkerd2 on 🌱 taru [📦📝🤷‍] via 🐼 v1.16.7 via  via ❄️  impure (shell) took 2s
➜ ./bin/go-run cli viz stat srv -A -owide                                                                                                         ~/work/linkerd2
NAMESPACE   NAME          UNAUTHORIZED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN   READ_BYTES/SEC   WRITE_BYTES/SEC
emojivoto   emoji-grpc          0.3rps   100.00%   1.1rps           0ms           0ms           0ms          1         156.5B/s         1282.4B/s
emojivoto   prom                0.0rps         -        -             -             -             -          -                -                 -
emojivoto   voting-grpc         0.0rps    87.88%   0.6rps           0ms           0ms           0ms          1          53.5B/s           31.5B/s
emojivoto   web-http            0.0rps    61.18%   1.4rps           1ms           2ms           2ms          1         110.2B/s         2195.7B/s

# after changing the `web-http` authz

on ⛵ kind-kind  linkerd2 on 🌱 taru [📦📝🤷‍] via 🐼 v1.16.7 via  via ❄️  impure (shell)
➜ ./bin/go-run cli viz stat srv -A -owide                                                                                                         ~/work/linkerd2
NAMESPACE   NAME          UNAUTHORIZED   SUCCESS   RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN   READ_BYTES/SEC   WRITE_BYTES/SEC
emojivoto   emoji-grpc          0.0rps         -     -             -             -             -          -                -                 -
emojivoto   prom                0.0rps         -     -             -             -             -          -                -                 -
emojivoto   voting-grpc         0.0rps         -     -             -             -             -          -                -                 -
emojivoto   web-http            1.0rps         -     -             -             -             -          -                -                 -



```

adleong reviewed (approved):
Looks great!  Nice work, @Pothulapati !!

alpeb reviewed (approved):
Looks pretty good to me! :+1: 
For your consideration in a future refactoring: in `stat_summary.go` we're starting to see separate groups of inter-related functions, so it might be worth separating them in different files under a new `stats` package...

Pothulapati commented:
> For your consideration in a future refactoring: in stat_summary.go we're starting to see separate groups of inter-related functions, so it might be worth separating them in different files under a new stats package...

Yep, Should make it easier to read!
